### PR TITLE
Fix problems with async code

### PR DESCRIPTION
**Pull request by implausible**
In asynchronous libgit calls with callbacks, when the callback fails or returns an error, the asynch isn't considered a failure. It also loops forever, even after the promise has rejected. The promise should then instead reject, return an error to the appropriate caller, and return that error to the javascript layer on nodegit as a rejection.

Now when a promise rejects in a callback, this should call the function calling the callback to reject as well.

### CODE REVIEW

**Comment by johnhaley81** [comment]
@implausible can you make another PR to update libgit2? This is way to huge to figure out what's going on.

**Comment by implausible** [comment]
See https://github.com/nodegit/nodegit/pull/743 for libgit2 bump

File: generate/templates/partials/field_accessors.cc [hunk 1]
```diff
@@ -258,7 +261,7 @@
         else {
           // promise was rejected
           baton->result = {{ field.return.error }};
-          baton->done = false;
+          baton->done = true;
```
Comment (johnhaley81) [inline generate/templates/partials/field_accessors.cc]: I don't remember why we did this originally but this change makes a lot more sense now. :+1:

**Comment by johnhaley81** [comment]
I'm good on this. Thanks @implausible !

**Pull request merged**